neqo-http3/src/response_stream.rs

```diff
@@ -0,0 +1,299 @@
+                return Err(Error::HttpFrameUnexpected);
+            }
+            ResponseStreamState::WaitingForData => {
+                if len > 0 {
+                    if fin {
+                        return Err(Error::HttpFrameError);
+                    }
+                    self.state = ResponseStreamState::ReadingData {
+                        remaining_data_len: len as usize,
+                    };
+                }
+            }
+            _ => unreachable!("This functions is only called in WaitingForResponseHeaders | WaitingForData | WaitingForFinAfterTrailers state.")
+        }
+        Ok(())
+    }
+
+    fn add_headers(&mut self, headers: Option<Vec<Header>>) -> Res<()> {
+        if self.response_headers_state != ResponseHeadersState::NoHeaders {
+            debug_assert!(
+                false,
+                "self.response_headers_state must be in state ResponseHeadersState::NoHeaders."
+            );
+            return Err(Error::HttpInternalError);
+        }
+        self.response_headers_state = ResponseHeadersState::Ready(headers);
+        self.conn_events.header_ready(self.stream_id);
+        self.state = ResponseStreamState::WaitingForData;
+        Ok(())
+    }
+
+    fn set_state_to_close_pending(&mut self) {
+        // Stream has received fin. Depending on headers state set header_ready
+        // or data_readable event so that app can pick up the fin.
+        qtrace!(
+            [self],
+            "set_state_to_close_pending:  response_headers_state={:?}",
+            self.response_headers_state
+        );
+        match self.response_headers_state {
+            ResponseHeadersState::NoHeaders => {
+                self.conn_events.header_ready(self.stream_id);
+                self.response_headers_state = ResponseHeadersState::Ready(None);
+            }
+            // In Ready state we are already waiting for app to pick up headers
+            // it can also pick up fin, so we do not need a new event.
+            ResponseHeadersState::Ready(..) => {}
+            ResponseHeadersState::Read => self.conn_events.data_readable(self.stream_id),
+        }
+        self.state = ResponseStreamState::ClosePending;
+    }
+
+    fn recv_frame(&mut self, conn: &mut Connection) -> Res<(Option<HFrame>, bool)> {
+        qtrace!([self], "receiving frame header");
+        let fin = self.frame_reader.receive(conn, self.stream_id)?;
+        if !self.frame_reader.done() {
+            Ok((None, fin))
+        } else {
+            qdebug!([self], "A new frame has been received.");
+            Ok((Some(self.frame_reader.get_frame()?), fin))
+        }
+    }
+
+    pub fn read_response_headers(&mut self) -> Res<(Vec<Header>, bool)> {
+        if let ResponseHeadersState::Ready(ref mut headers) = self.response_headers_state {
+            let mut tmp = Vec::new();
+            if let Some(ref mut hdrs) = headers {
```

Comment:
You shouldn't use this pattern.  Try:

```rust
let tmp = headers.unwrap_or_else(|| Vec::new());
```

Try to think of a better name than `tmp` too.